File: 0823. Binary Trees With Factors.java
class Solution {
    final int MOD = 1000000007;
    public int numFactoredBinaryTrees(int[] A) {
        int N = A.length;
        Arrays.sort(A);
        long[] dp = new long[N];
        Arrays.fill(dp, 1); //each can be a tree by itself
        HashMap < Integer, Integer > map = new HashMap < > ();
        for (int i = 0; i != N; i++) {
            map.put(A[i], i); //maps value to index
            for (int j = 0; j != i; j++) { //no A[i]=1, so n=n*1 can't be a binary tree
                if (A[i] % A[j] == 0 && map.containsKey(A[i] / A[j])) //factor pair found
                    dp[i] = (dp[i] + dp[j] * dp[map.get(A[i] / A[j])]) % MOD; 
                    //add #ways to make factor1 * #ways to make factor2 to dp[i], then mod it
            }
        }
        long res = 0;
        for (long d: dp) res += d; //<=1000 ints, so can't go over long's max
        return (int)(res % MOD);
    }
}

File: 0823. Binary Trees With Factors.cpp
class Solution {
    public:
        int numFactoredBinaryTrees(vector < int > & A) {
            sort(A.begin(), A.end());
            unordered_map < int, long > dp;
            unordered_set < int > s;

            for (int a: A) {
                s.insert(a);
                dp[a]++; //one tree of each num by itself
            }

            long total = 1;
            int mod = 1e9 + 7;
            vector < int > nums = A;
            for (int i = 1; i < A.size(); i++) {
                for (int j = 0; j < i; j++) { //search thru all smaller nums
                    if (nums[i] % nums[j] == 0 && s.find(nums[i] / nums[j]) != s.end()) { //if we found a factor and its partner
                        dp[nums[i]] += dp[nums[j]] * dp[nums[i] / nums[j]]; //#ways to make LHS * #ways to make RHS
                        dp[nums[i]] %= mod;
                    }
                }
                total += dp[nums[i]];
                total %= mod;
            }
            return static_cast < int > (total);
        }
};
